### PR TITLE
AWS Security Group Rules properties and matchers

### LINKED ISSUE #2574: Add ability to query for ingress rules to security group resources
<a href="https://github.com/clintoncwolfe"></img></a> **Issue by [clintoncwolfe](https://github.com/clintoncwolfe)**
_Thursday Dec 14, 2017 at 15:47 GMT_
_Originally opened as https://github.com/chef/inspec-aws/issues/137_

----

I'd like to be able to do something like:
```
describe aws_ec2_security_group('sg-12345678') do
  its('ingress_rule.count') { should be 2 }
  its('ingress_rules.last') { should include(:action => 'REJECT', :source =>'0.0.0.0/0') }
  it { should_not be_open_to_the_world_on_port(22) }
  its('ingress_rules.first') { should include(source: 'sg-87654321', port_from: 22, port_to: 22) }
end
```

So, for the singular resource, I was thinking of having the `ingress_rules` property be an array of hashes, sorted by rule order.  Based on that property, I'd also add a high-level matcher, `open_to_the_world_on_port?(N)`.

And for the plural case:
```
describe aws_ec2_security_groups.where { open_to_the_world_on_port(22) }  do
  it { should_not exist }
end
```

And perhaps:
```
describe aws_ec2_security_groups.where { blocks_all_ingress }  do
  it { should exist }
end
```

Refs #129

### PR DESCRIPTION
**Pull request by clintoncwolfe**
Provides low-, and mid-level properties and matchers for examining rules on `aws_security_group`.  Fixes #2574.

Low-level: `inbound_rules` and `outbound_rules` properties provide access to the arrays-of-structs provided by the AWS API.

Mid-level: `allow_in`, `allow_in_only`, `allow_out`, and `allow_out_only` provide an expressive way of interrogating the rulesets without navigating the structs provided by the API. Documentation on these is comprehensive; I'm especially interested in feedback regarding clarity of these docs; misinterpretation is especially concerning here.

TODO: The output formatting is not what I would hope for on the `allow_*` matchers (it always shows an empty hash, instead of the actual criteria used).  I think that is a general reporting issue, not related specifically to this feature.  

Integration and unit test run screenshots attached below (note: the screenshots show a high-level matcher, `open_to_the_world`, which has since been removed.)

screen shot 2018-03-24 at 12 46 19
screen shot 2018-03-24 at 12 47 46

### CODE REVIEW

File: docs/resources/aws_security_group.md.erb
Comment (burtlo) [inline docs/resources/aws_security_group.md.erb]: I would write Security Group (SG) so you can use the SG abbreviation with wild abandon after that.

File: docs/resources/aws_security_group.md.erb
Comment (burtlo) [inline docs/resources/aws_security_group.md.erb]: The sentence above is talking about a security group that cannot be found. This example is describing a look at the first rule. It may be useful to include an example that is tied to the sentence above.

The `linux_servers` variable is not defined and could create confusion.

File: docs/resources/aws_security_group.md.erb
Comment (burtlo) [inline docs/resources/aws_security_group.md.erb]: Same notes as previous about tying the words above with the code sample below. `isolated_servers` is not defined.

File: docs/resources/aws_security_group.md.erb
Comment (burtlo) [inline docs/resources/aws_security_group.md.erb]: Most likely will need to deprecate `be_open_to_the_world` when aliens land we take more of a `universe` perspective. 😜

**Review by burtlo — COMMENTED** [review]
Documentation review.

File: lib/resources/aws/aws_security_group.rb
Comment (burtlo) [inline lib/resources/aws/aws_security_group.rb]: I would suggest displaying all of the remaining `raw_criteria.keys` when they are present.

File: lib/resources/aws/aws_security_group.rb
Comment (burtlo) [inline lib/resources/aws/aws_security_group.rb:64]: I would consider `expressed_criteria` as it is the criteria expressed in the test.

File: lib/resources/aws/aws_security_group.rb
Comment (burtlo) [inline lib/resources/aws/aws_security_group.rb]: I find conditionals that filters inside of iteration difficult to recognize while reading the code. Perhaps consider filtering first and then building the `recognized_criteria`. Also proposed is a way to remove the empty hash accumulator that is created before the iteration and assign it with the structure that is built from the filtering.

```
supported_criteria = allowed_criteria.find_all { |criterion| expressed_critiera.key?(criterion) }

recognized_criteria = supported_criteria.inject({}) do |recognized,criterion|
  recognized.merge(criterion => raw_criteria.delete(criterion))
end
```

File: lib/resources/aws/aws_security_group.rb
Comment (burtlo) [inline lib/resources/aws/aws_security_group.rb]: Some teams I have worked with have considered an early return in cases like these. It raises the cyclomatic complexity. Similar to your `allow_match_protocol` defined later.

File: lib/resources/aws/aws_security_group.rb
Comment (burtlo) [inline lib/resources/aws/aws_security_group.rb]: Should this allow 'first' and 'last' strings as well?

File: lib/resources/aws/aws_security_group.rb
Comment (burtlo) [inline lib/resources/aws/aws_security_group.rb:155]: `Array(query)` should convert a non-Array to an Array and leave a variable that is already an Array alone.

**Review by burtlo — COMMENTED** [review]
A few suggestions around the code.

File: docs/resources/aws_security_group.md.erb
Comment (kagarmoe) [reply docs/resources/aws_security_group.md.erb]: While writing SG is certainly faster--and reading 'security group' over and over gets tedious--this type of reference documentation is necessarily designed to be dipped into. It is infrequently read from top-to-bottom. So, while I agree that we should be consistent in either 'Security Group' or 'SG', I prefer the former because it is less jarring for users. (I haven't compared to other docs, but I suspect we've generally used the former.)

Scratch all of that -- you've handled it well in the body by starting out with "Security Group" and then following with SG. Neatly solved!

File: docs/resources/aws_security_group.md.erb
Comment (kagarmoe) [inline docs/resources/aws_security_group.md.erb]: 'certain' in this context is ambiguous. Do you mean "specific"?

**Review by kagarmoe — APPROVED** [review]

File: docs/resources/aws_security_group.md.erb
Comment (clintoncwolfe) [reply docs/resources/aws_security_group.md.erb]: Yes, specific is intended.  Thanks!

File: lib/resources/aws/aws_security_group.rb
Comment (clintoncwolfe) [reply lib/resources/aws/aws_security_group.rb]: I'm writing for maintainability and readability; I appreciate the feedback.
I don't think injection and merge is _more_ clear, however.  Many junior devs struggle with inject.

I'll move the `if` to be a block conditional, instead of an inline conditional, so it stands out.

**Comment by clintoncwolfe** [comment]
@burtlo, do you have any feedback about the UX of this feature?

File: lib/resources/aws/aws_security_group.rb
Comment (burtlo) [inline lib/resources/aws/aws_security_group.rb:89]: I assume that double underscore was intentional?

File: docs/resources/aws_security_group.md.erb
Comment (burtlo) [inline docs/resources/aws_security_group.md.erb]: Space is needed on the backside of the array.

File: docs/resources/aws_security_group.md.erb
Comment (burtlo) [inline docs/resources/aws_security_group.md.erb]: No spacing at all here and there is spacing above. What's the standard for the InSpec team?

**Comment by burtlo** [comment]
My only feedback on the experience is on  `be_open_to_the_world`.

* AWS documentation on security groups [documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-network-security.html) doesn't mention 'world'.  From skimming through it, it may be more appropriate to use `be_open_to_all_traffic`.

* Ask a non-native English speaker if that is clear.

**Comment by clintoncwolfe** [comment]
Good idea, I've asked about `be_open_to_the_world` in [community slack](https://chefcommunity.slack.com/archives/C1XB6U6MN/p1522341680000701).

File: docs/resources/aws_security_group.md.erb
Comment (chris-rock) [inline docs/resources/aws_security_group.md.erb]: Let's not use implicit matcher. With our move from ServerSpec to InSpec, we decided to reduce the amount of matcher. They hide logic and it is not obvious to the reader what this would mean. Already, in this case you are limiting the matcher to ipv4. Which would be immediatate clear if a user would write:

```
it { should_not allow_in(ipv4_range: '0.0.0.0/0')}
vs
it { should_not be_open_to_the_world }
```
I think the limited amount of chars that you safe does not justify the use of a custom matcher.

**Review by chris-rock — DISMISSED** [review]
Thank you @clintoncwolfe for this improvement. I think we should not use `open_to_the_world` and `open_to_the_world_on_port?`

**Comment by clintoncwolfe** [comment]
@chris-rock, I've removed `open_to_the_world`.  Are there any other changes you'd like to see before you can approve this?  Thanks!

**Review by jquick — APPROVED** [review]
This looks good to me! Thanks @clintoncwolfe

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: Could we get a space between `are` and `nil`

Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: I notice we capitalize `Security Group` some places but not in others, could we standardize?

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: Could we add an ending punctuation?

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: Capitalize `this`? I don't particularly mind, but looks the the others have the first letter capitalized.

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: Could we add a space before the ending `}`?

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: No need for a comma here.

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: Space before ending `}`?

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: I seem to remember someone ( @chris-rock? ) mention avoiding chaining methods in the `its` block...I don't know how to avoid it here though.

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: Could we change `levle` to `level`?

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: Is there a leading space here?

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: Could we add a space before ending `}` on the examples below?

File: docs/resources/aws_security_group.md.erb
Comment (jerryaldrichiii) [inline docs/resources/aws_security_group.md.erb]: Could we change this to `Ensure that no IPs are allowed in`?

File: lib/resources/aws/aws_security_group.rb
Comment (jerryaldrichiii) [inline lib/resources/aws/aws_security_group.rb]: Single quotes are not required around non-interpolated strings...but I like them. I'll leave it up to you to change it.

**Review by jerryaldrichiii — CHANGES_REQUESTED** [review]
Superb work here @clintoncwolfe! 100% of my requests are nit-picks.

The implementation is fantastic and provides a ton of value.

File: docs/resources/aws_security_group.md.erb
Comment (clintoncwolfe) [reply docs/resources/aws_security_group.md.erb]: We document this pattern on [the glossary page](https://www.inspec.io/docs/reference/glossary/), section Advanced Syntax.

If you prefer, we could add `inbound_rules_count` and `outbound_rules_count` properties.

**Review by jerryaldrichiii — APPROVED** [review]
Great work @clintoncwolfe !

**Comment by Sigmonia** [comment]
Loving the new infrastructure controls.

Is there a (feature/enhancement) issue tracking support for inbound rules using security groups as the traffic source?

**Pull request merged**
